### PR TITLE
simple: adds S1031 LintNilCheckAroundRange

### PR DESCRIPTION
**Pull request by scriptnull**
Fixes https://github.com/dominikh/go-tools/issues/67

I am relatively new to manipulating AST and this code base. So, please take a look at the code and let me know if I have made some naive mistakes.

Thank you.

### CODE REVIEW

File: simple/lint.go [hunk 1]
```diff
@@ -68,6 +68,43 @@ func (c *Checker) Funcs() map[string]lint.Func {
 		"S1028": c.LintErrorsNewSprintf,
 		"S1029": c.LintRangeStringRunes,
 		"S1030": c.LintBytesBufferConversions,
+		"S1031": c.LintNilCheckAroundRange,
+	}
+}
+
+func (c *Checker) LintNilCheckAroundRange(j *lint.Job) {
```
Comment (dominikh) [inline simple/lint.go]: This check is missing one additional condition: We can only suggest the simplification if ranging over a slice.

In the case of channels, ranging over a nil channel would block forever, which is why the check is needed.

In the case of maps, the loop body may try to insert values in the map. Trying to do that on a nil map panics.

File: simple/testdata/LintNilCheckAroundRange.go [hunk 2]
```diff
@@ -0,0 +1,11 @@
+package pkg
+
+func fn() {
+	str := []string{}
+
+	if str != nil { // MATCH /unnecessary nil check around range/
```
Comment (dominikh) [inline simple/testdata/LintNilCheckAroundRange.go]: Add a couple more test cases, namely:

- a range with no check around it (shouldn't match)
- a check with more statements than just the range (shouldn't match)
- ranging over channels (shouldn't match)
- ranging over maps (shouldn't match)

File: simple/lint.go (hunk 1 above)
Comment (dominikh) [inline simple/lint.go]: Forget my silly comment regarding nil maps. Ranging over a nil map will never execute the loop body. So, we should check that the type is not a channel, slices and maps are fine.

File: simple/testdata/LintNilCheckAroundRange.go (hunk 2 above)
Comment (dominikh) [inline simple/testdata/LintNilCheckAroundRange.go]: Based on my comment above, ranging over maps **should** match.

**Comment by scriptnull** [comment]
@dominikh Thanks for the review.

I think, we need to also match for nil check around range over strings

```go
	var emptyStr string
	if emptyStr != "" { 
		for _, x := range emptyStr {
			fmt.Println(x)
		}
	}
```
is same as
```go
	var emptyStr string
	for _, x := range emptyStr {
		fmt.Println(x)
	}
```

can I add this case also in this linting rule itself ?

**Comment by dominikh** [comment]
I think the string case is contrived and not worth adding.

**Comment by scriptnull** [comment]
Alright, I will not implement that case then.

**Comment by scriptnull** [comment]
@dominikh I have updated the code. Let me know if you need something more. Thanks.

**Comment by dominikh** [comment]
Merged as 7c33446 and 2ab10e4, thank you.

**Pull request closed without merging**